src/planner/binder/statement/bind_drop.cpp

```diff
@@ -43,10 +46,32 @@ BoundStatement Binder::Bind(DropStatement &stmt) {
 		stmt.info->schema = entry->schema->name;
 		break;
 	}
-	case CatalogType::DATABASE_ENTRY:
-		// attaching and detaching is read-only
-		stmt.info->catalog = SYSTEM_CATALOG;
+	case CatalogType::DATABASE_ENTRY: {
+		// use DuckDB default impl if no storage extension is registered to handle this functionality
+		auto &base = (DropInfo &)*stmt.info;
+		string database_name = base.name;
+
+		auto &config = DBConfig::GetConfig(context);
+
+		if (config.storage_extensions.empty()) {
+			// attaching and detaching is read-only
+			stmt.info->catalog = SYSTEM_CATALOG;
+		}
+		// for now assume only one storage extension provides the custom drop_database impl
+		for (auto &extension_entry : config.storage_extensions) {
+			if (extension_entry.second->create_database != nullptr) {
```

Comment:
Generally we prefer early-out to reduce the indentation depth and increase readability, e.g.:

```sql
if (extension_entry.second-> drop_database == nullptr) {
    continue;
}
...
```